src/aleph/http/client.clj

```diff
@@ -386,95 +392,314 @@
+     max-chunk-size          65536
+     idle-timeout            0}}]
+  (let [handler (if raw-stream?
+                  (http1-raw-client-handler responses response-buffer-size)
+                  (http1-client-handler responses response-buffer-size))]
+
+    (-> pipeline
+        (common/attach-idle-handlers idle-timeout)
         (.addLast "http-client"
-          (HttpClientCodec.
-            max-initial-line-length
-            max-header-size
-            max-chunk-size
-            false
-            false))
+                  (HttpClientCodec.
+                    max-initial-line-length
+                    max-header-size
+                    max-chunk-size
+                    false
+                    false))
         (.addLast "streamer" ^ChannelHandler (ChunkedWriteHandler.))
         (.addLast "handler" ^ChannelHandler handler)
-        (http/attach-idle-handlers idle-timeout))
-      (when (some? proxy-options)
-        (let [proxy (proxy-handler (assoc proxy-options :ssl? ssl?))]
-          (.addFirst pipeline "proxy" ^ChannelHandler proxy)
-          ;; well, we need to wait before the proxy responded with
-          ;; HTTP/1.1 200 Connection established
-          ;; before sending any requests
-          (when (instance? ProxyHandler proxy)
-            (.addAfter pipeline
-              "proxy"
-              "pending-proxy-connection"
+        (add-proxy-handlers responses proxy-options ssl?)
+        (add-non-http-handlers logger pipeline-transform))))
+
+
+(defn http2-client-handler
+  "Given a response-stream, returns a ChannelInboundHandler that processes
+   inbound Netty Http2 frames, converts them, and places them on the stream"
+  [^Http2StreamChannel ch response-d raw-stream? ^long buffer-capacity]
+  (let [complete (d/deferred) ; realized when this stream is done, regardless of success/error
+
+        ;; if raw, we're returning ByteBufs, if not, we convert to byte[]'s
+        body-stream
+        (doto (if raw-stream?
+                (netty/buffered-source ch #(.readableBytes ^ByteBuf %) buffer-capacity)
+                (netty/buffered-source ch #(alength ^bytes %) buffer-capacity))
+              (s/on-closed #(d/success! complete true)))
+
+        handle-error
+        (let [close-body-handler (fn [_] (s/close! body-stream))]
+          (fn handle-error [ex]
+            (log/error ex "Exception caught in HTTP/2 stream client handler" {:raw-stream? raw-stream?})
+            ;; if we've already returned a response, best we can do is blow up the
+            ;; body stream and the complete deferred
+            (if-not (d/realized? response-d)
+              (d/error! response-d ex)
+              (do
+                (d/error! complete ex)
+                (-> (s/put! body-stream ex)
+                    (d/on-realized close-body-handler close-body-handler))))))
+
+        handle-shutdown-frame
+        (fn handle-shutdown-frame [evt]
+          (when (or (instance? Http2ResetFrame evt)
```

Comment:
Sad those two classses are not sharing a common interface.

Reply:
Yeah...but they're meant for different use cases. From the perspective of a stream handler, they both mean "shut down this stream", but GOAWAY is for the whole connection, and RST_STREAM is just for this stream.